# AG-HYPOPT · Trial 01

> ▶ RUN = execute this cell, never modify it  ·  ✍️ WRITE = replace the [placeholder] with your text  ·  ✍️ SET = put your value where the [placeholder] is, then run the cell. Full rules: `instructions.md`


## 1. ✍️ Read and summarize

**MACHINERY TEST (clean branch, test phase).** Campaign restarted on the
simplified stack: `objective(params) -> (loss, uncertainty, report)` over the
frozen vanilla benchmark (cheap test budget n_runs=n_iter=100, ~1h/trial), 5
tunables (sigma_ref, lr_mu, lr_gamma, gamma_anneal, clip). Registry: baseline_17g
only (anchor at 200x200 — absolute numbers not yet comparable). Purpose of this
trial: exercise the packed objective + the loop end-to-end. Uniform phase (n<5):
batch is fully random; candidate picked closest to vanilla defaults.


In [1]:
# 2. ▶ RUN - Propose candidate trials (AGHyperopt)
TRIAL_ID = 'trial_01'   # auto-stamped by the generator; do not edit
import os

EXPERIMENT_DIR = os.getcwd()
SPACE_PATH = os.path.join(EXPERIMENT_DIR, 'space.json')
TRIALS_PATH = os.path.join(EXPERIMENT_DIR, 'trials.json')

MAX_TRIALS = 10    # campaign cap: stop generating new trials after this many (None = unlimited)

from ag_hypopt import AGHyperopt

opt = AGHyperopt()
opt.fit(SPACE_PATH, TRIALS_PATH)
proposed_trials = opt.propose_trials(10)


phase = uniform  (1/5 trials)
ID |     ei | params
 1 |      - | {"clip": 7.731758691196147, "gamma_anneal": 0.5089869401329147, "lr_gamma": 0.2364422836258463, "lr_mu": 23.667102605821455, "sigma_ref": 22.592978164009814}
 2 |      - | {"clip": 7.527715206818916, "gamma_anneal": 0.11434220214494706, "lr_gamma": 0.4458022622392636, "lr_mu": 29.934739660952516, "sigma_ref": 17.72832883159676}
 3 |      - | {"clip": 18.65210787690615, "gamma_anneal": 0.24115931782756445, "lr_gamma": 0.8977344490032726, "lr_mu": 20.466463397515113, "sigma_ref": 22.152407879894366}
 4 |      - | {"clip": 6.247760360260092, "gamma_anneal": 0.5788158108647133, "lr_gamma": 0.650093094592597, "lr_mu": 22.742081065538873, "sigma_ref": 12.249936502126333}
 5 |      - | {"clip": 12.706705329053676, "gamma_anneal": 0.3179513179030421, "lr_gamma": 0.8614242219589323, "lr_mu": 22.89899089810529, "sigma_ref": 7.200320862727764}
 6 |      - | {"clip": 7.247574146787341, "gamma_anneal": 0.6567870521895803, "lr_gamma": 

## 3. ✍️ Analyze and choose

**Uniform phase (machinery test)** — the batch is fully random by design
(n=1 completed < n_initial=5). No physics claim is made at this stage: the
trial exists to exercise `objective()` and the loop. Picking the candidate
closest to the vanilla defaults (least normalized deviation), ID 4:
{"clip": 6.247760360260092, "gamma_anneal": 0.5788158108647133, "lr_gamma": 0.650093094592597, "lr_mu": 22.742081065538873, "sigma_ref": 12.249936502126333}. Expect a loss in the baseline ballpark once re-anchored
at the 100x100 budget.


In [2]:
# 4. ✍️ SET - Chosen candidate (edit INDEX below, then run this cell)
INDEX = 4            # [Put the index here: 1-based candidate number from the table in cell 2]


In [3]:
# 5. ▶ RUN - Execute the chosen trial  (do not interrupt unless obviously broken)
assert INDEX is not None, 'INDEX not set: edit the SET cell (cell 4) before running this one'
assert 1 <= INDEX <= len(proposed_trials), f'bad INDEX: {INDEX}'
CHOSEN = proposed_trials[INDEX - 1]['params']
print('chosen:', CHOSEN)

from ag_hypopt import objective

import traceback, time
t0 = time.time()
try:
    loss, uncertainty, report = objective(CHOSEN)
    print(f'trial finished in {(time.time()-t0)/60:.1f} min')
    print(f'objective (MSE vs true) = {loss:.4f} ± {uncertainty:.4f}')
    print(report)
except Exception:
    traceback.print_exc()
    loss, uncertainty = None, None


chosen: {'clip': 6.247760360260092, 'gamma_anneal': 0.5788158108647133, 'lr_gamma': 0.650093094592597, 'lr_mu': 22.742081065538873, 'sigma_ref': 12.249936502126333}
Running  1nW Trans05 ... 

μ 9.39 -> 8.67 | γ 8.5 -> 10.07 | NLL 1.19


Running  1nW Trans10 ... 

μ 12.37 -> 12.67 | γ 8.5 -> 8.37 | NLL 1.79


Running  1nW Trans20 ... 

μ 17.32 -> 17.44 | γ 8.5 -> 8.77 | NLL 2.59


Running  1nW Trans40 ... 

μ 38.41 -> 38.65 | γ 8.5 -> 8.63 | NLL 3.21


Running  1nW Trans60 ... 

μ 61.37 -> 61.63 | γ 8.5 -> 8.61 | NLL 2.95


Running  1nW Trans80 ... 

μ 79.36 -> 79.41 | γ 8.5 -> 8.60 | NLL 2.63


Running 1nW Trans100 ... 

μ 70.82 -> 72.01 | γ 8.5 -> 8.57 | NLL 3.38


Running  3nW Trans05 ... 

μ 13.20 -> 11.52 | γ 14.1 -> 14.57 | NLL 1.55


Running  3nW Trans10 ... 

μ 24.48 -> 24.61 | γ 14.1 -> 14.18 | NLL 2.59


Running  3nW Trans20 ... 

μ 34.28 -> 34.70 | γ 14.1 -> 13.96 | NLL 3.11


Running  3nW Trans40 ... 

μ 84.89 -> 87.02 | γ 14.1 -> 14.16 | NLL 2.61


Running  3nW Trans60 ... 

μ 103.20 -> 103.62 | γ 14.1 -> 14.16 | NLL 3.46


Running  3nW Trans80 ... 

μ 137.54 -> 137.05 | γ 14.1 -> 14.18 | NLL 3.15


Running 3nW Trans100 ... 

μ 175.71 -> 177.85 | γ 14.1 -> 14.18 | NLL 2.94



Total: 52.8 min
trial finished in 52.8 min
objective (MSE vs true) = 0.0022 ± 0.0013
exp           μ_true   μ_fit     Δμ% |   γ_true   γ_fit     Δγ%
1nW Trans05     9.39    8.67    -7.7 |     8.50   10.07   +18.5
1nW Trans10    12.37   12.67    +2.4 |     8.50    8.37    -1.5
1nW Trans20    17.32   17.44    +0.7 |     8.50    8.77    +3.1
1nW Trans40    38.41   38.65    +0.6 |     8.50    8.63    +1.6
1nW Trans60    61.37   61.63    +0.4 |     8.50    8.61    +1.3
1nW Trans80    79.36   79.41    +0.1 |     8.50    8.60    +1.2
1nW Trans100   70.82   72.01    +1.7 |     8.50    8.57    +0.8
3nW Trans05    13.20   11.52   -12.7 |    14.10   14.57    +3.3
3nW Trans10    24.48   24.61    +0.6 |    14.10   14.18    +0.6
3nW Trans20    34.28   34.70    +1.2 |    14.10   13.96    -1.0
3nW Trans40    84.89   87.02    +2.5 |    14.10   14.16    +0.4
3nW Trans60   103.20  103.62    +0.4 |    14.10   14.16    +0.4
3nW Trans80   137.54  137.05    -0.4 |    14.10   14.18    +0.5
3nW Trans100  175.

## 6. ✍️ WRITE - Analyze the results

**Machinery test - verdict: PASS.** The packed `objective(params) -> (loss, uncertainty, report)`
executed end-to-end on the frozen vanilla benchmark at the cheap 100x100 test budget
(14 exps: 1nW & 3nW x 7 transfer points, 52.8 min wall), returning
`MSE 0.0022 ± 0.0013` plus a clean per-experiment report - no crashes, no stalled runs.

Per-experiment pattern: μ tracks well across both systems (rel-RMSE 4.2%, RMSE 1.026) with a
slight positive bias (+0.322); γ has smaller absolute RMSE (0.453) but larger relative error
(5.2%, bias +0.201). Deviations concentrate at the low-transfer endpoint (Trans05):
3nW Trans05 μ is the worst outlier (-12.7%), 1nW Trans05 shows μ -7.7% and γ +18.5% (the worst
γ outlier); every higher transfer point sits within ~±3%. The chosen candidate (closest to
vanilla defaults, ID 4) behaved as expected.

Caveat: loss 0.0022 is NOT comparable to registry baseline_17g (0.001578 at the 200x200 anchor)
- different budget, so no improvement claim is made from this trial.

What the test proved: the objective is properly packed (loss, uncertainty and report from one
call), the trial loop (propose -> choose -> run -> record -> generate next) works on the clean
branch, and the RUN/SET/WRITE notebook protocol survives an unattended ~1h run.

**Summary:** Machinery test PASS at the 100x100 budget: MSE 0.0022 ± 0.0013, μ/γ rel-RMSE 4.2%/5.2%, only the Trans05 endpoints outlying; loss not comparable to the 200x200 baseline_17g anchor yet.
**Key insight:** The simplified packed objective + trial loop is proven end-to-end and loop-ready; loss numbers only become meaningful once re-anchored at a matched budget.

In [4]:
# 7. ✍️ SET - Summary and key insight for the record (edit the strings below, then run)
SUMMARY = 'Machinery test PASS at the 100x100 budget: MSE 0.0022 ± 0.0013, μ/γ rel-RMSE 4.2%/5.2%, only the Trans05 endpoints outlying; loss not comparable to the 200x200 baseline_17g anchor yet.'
KEY_INSIGHT = 'The simplified packed objective + trial loop is proven end-to-end and loop-ready; loss numbers only become meaningful once re-anchored at a matched budget.'


In [5]:
# 8. ▶ RUN - Record the trial in trials.json
import json

assert SUMMARY is not None, 'SUMMARY not set: fill the SET cell (cell 7) first'
assert KEY_INSIGHT is not None, 'KEY_INSIGHT not set: fill the SET cell (cell 7) first'

entry = {
    'trial_id': TRIAL_ID,
    'config': CHOSEN,
    'objective': loss,
    'uncertainty': uncertainty,
    'summary': SUMMARY,
    'key_insight': KEY_INSIGHT,
    'notebook': f'{TRIAL_ID}.ipynb',
}
data = json.load(open(TRIALS_PATH))
assert not any(t.get('trial_id') == TRIAL_ID for t in data['trials']),     f'{TRIAL_ID} is already registered in trials.json'
data['trials'].append(entry)
json.dump(data, open(TRIALS_PATH, 'w'), indent=2)
best = min((t for t in data['trials'] if t.get('objective') is not None),
           key=lambda t: t['objective'], default=None)
print('saved', TRIAL_ID, '| current best:', best['trial_id'] if best else None)


saved trial_01 | current best: baseline_17g


In [6]:
# 9. ▶ RUN - Generate the next trial (or end the campaign)
import os, re, json as _json

num = int(re.search(r'(\d+)$', TRIAL_ID).group(1))
nxt = num + 1
if MAX_TRIALS is not None and nxt > MAX_TRIALS:
    print(f'Campaign complete: cap MAX_TRIALS={MAX_TRIALS} reached after {TRIAL_ID}. Stop here.')
else:
    template_path = os.path.join(EXPERIMENT_DIR, 'template.ipynb')
    target = os.path.join(EXPERIMENT_DIR, f'trial_{nxt:02d}.ipynb')
    assert not os.path.exists(target), f'{target} already exists: refusing to overwrite'

    nb = _json.load(open(template_path))

    def _stamp(old, new):
        for c in nb['cells']:
            if old in ''.join(c.get('source', [])):
                c['source'] = [s.replace(old, new) for s in c['source']]
                return True
        raise RuntimeError(f'stamp target {old!r} not found in the template')

    _stamp('{N}', f'{nxt:02d}')
    _stamp('trial_XXX', f'trial_{nxt:02d}')
    _json.dump(nb, open(target, 'w'), indent=1)
    print(f'Created {os.path.basename(target)}. Open it and follow its cells from the top.')


Created trial_02.ipynb. Open it and follow its cells from the top.
